Add tests for branch operators + Bugfix for conditional branch label address parsing

dezuman opened the pull request.

dezuman committed:
Expand testing macros to allow asserting execution errors and more

dezuman committed:
Add helpful address constants for clarity (Avoiding "magic numbers")

dezuman committed:
Add WIP testing for branch operators

dezuman committed:
Do forgotten rustfmt

dezuman committed:
Fix missing bitshift for jump address in B<C> (Conditional branch)

Before it didn't actually hit the label location in the asm. Unconditional branching was parsing it correctly all along.

dezuman committed:
Fix state multiwriting macro to handle new non-option values

dezuman committed:
Finalize testing for branch operators

dezuman committed:
Fix comment

Earlz commented on tests/test_branch.rs:
```diff
@@ -0,0 +1,415 @@
+    // Common pre-execution state
+    common_state!(
+        applicable_op_ids,
+        vm_states.r[1] = Some(ASM_ENTRY + 2 * OP_SIZE + THUMBS_MODE)
```
This could likely be made less noisy by computing this address with a function like `test_code_offset(2 * OP_SIZE)` where code_offset is something like
```
fn test_code_offset(offset: u32) -> u32{
    ASM_ENTRY + THUMBS_MODE + offset
}
```

A simple change, but even with constants it's quite noisey about why THUMBS_MODE is needed and what ASM_ENTRY is

dezuman replied:
Good point, I'll do that! Probably as a macro though, I've walked too far down that path to turn back now.

Earlz replied:
Just be careful not to overuse macros where they aren't needed

dezuman replied:
Yeah, I'm working on when to use what.

Earlz commented on tests/test_branch.rs:
```diff
@@ -0,0 +1,415 @@
+
+    // 0: B <label> T2"
+    let ops0 = format!("b test1 {}", post_ops);
+    create_vm!(vms, vm_states, 0, code_var = true, ops0);
```
What is `code_var = true`? Looking through the macro it's not obvious what the purpose is and the syntax is a bit foreign here

dezuman replied:
Macro syntax is still very much up in the air. That line is really just a token to help both the test writer and the macro to know what macro pattern to use. I've intended to come up with something smarter anyway though, so I'll look into it.

dezuman committed:
Add functions to easily and clearly define memory addresses

dezuman committed:
Rename boilerplate variable to be more descriptive and less verbose

applicable_op_ids -> ops_to_test

It's just a vector that tell the macros which opcodes are being tested in each test function.

dezuman committed:
Reformat input of create_vm macro to be less obscure

dezuman committed:
Reformat input of run_tests macro to align with create_vm

dezuman committed:
Reformat input to state multisetting macro to be less obscure

dezuman commented:
Ok, enough reformatting now. Is this good to merge?

dezuman committed:
Fix sneak error caused by reformatting

dezuman commented:
Ok, since the last commits are purely visual I'll go ahead and merge
